KAFKA-10316: Consider renaming getter method for Interactive Queries

johnthotekat opened the pull request:
*Deprecated the existing getters and added new getters `activeHost()`, `standbyHosts()`, and `partition()` in KeyQueryMetadata.java.

*Updated the existing Tests to use the new getters of KeyQueryMetadata.java . Below are the Test classes that were updated with new getters.

    OptimizedKTableIntegrationTest.java
    QueryableStateIntegrationTest.java
    StoreQueryIntegrationTest.java
    StreamsMetadataStateTest.java

*Corrected JavaDoc nits in the deprecated methods.

@mjsax @abbccdda @vvcephei @brary Since you've voted on the KIP, please feel free to review. 

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

Commit:
KAFKA-10316: Consider renaming getter method for Interactive Queries

* Deprecated the existing getters and added new getters `activeHost()`, `standbyHosts()`, and `partition()` to KeyQueryMetadata.java.

* Update the existing Tests to use the new getters of KeyQueryMetadata.java . Below are the Test classes that were updated with new getters.

    OptimizedKTableIntegrationTest.java
    QueryableStateIntegrationTest.java
    StoreQueryIntegrationTest.java
    StreamsMetadataStateTest.java

mjsax commented on streams/src/main/java/org/apache/kafka/streams/KeyQueryMetadata.java:
```diff
@@ -71,11 +75,40 @@ public HostInfo getActiveHost() {
     }
 
+    /**
+     * Get the Active streams instance for given key
```
Nit: `[a]ctive` (i understand that it's just c&p but it seem wrong in the original JavaDoc, too.

`streams` -> `Kafka Streams`

Also, missing `.` at the end of the sentence.

johnthotekat replied:
I'm with you, we should fix it, for the deprecated ones too.

Fixed.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/KeyQueryMetadata.java:
```diff
@@ -71,11 +75,40 @@ public HostInfo getActiveHost() {
+    }
+
+    /**
+     * Get the Streams instances that host the key as standbys
```
nit: `Streams` -> `Kafka Streams`

Also, missing `.` at the end of the sentence.

johnthotekat replied:
Fixed.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/KeyQueryMetadata.java:
```diff
@@ -71,11 +75,40 @@ public HostInfo getActiveHost() {
+    }
+
+    /**
+     * Get the Store partition corresponding to the key.
```
`[s]tore`

johnthotekat replied:
Fixed.

mjsax reviewed (commented):
Thanks for the PR @johnthotekat!

Just some JavaDocs nits. I understand it's c&p but it seem the original JavaDocs are not "ideal" and we should just improve it -- if you want, you could also update the JavaDocs of the deprecated methods accordingly (I leave it up to you if you want to do this or not -- I am fine either way).

mjsax commented:
Retest this please.

Commit:
KAFKA-10316: Consider renaming getter method for Interactive Queries

* Deprecated the existing getters and added new getters `activeHost()`, `standbyHosts()`, and `partition()` to KeyQueryMetadata.java.

* Update the existing Tests to use the new getters of KeyQueryMetadata.java . Below are the Test classes that were updated with new getters.

    OptimizedKTableIntegrationTest.java
    QueryableStateIntegrationTest.java
    StoreQueryIntegrationTest.java
    StreamsMetadataStateTest.java

* Correcting some JavaDoc nits.

brary commented on streams/src/main/java/org/apache/kafka/streams/KeyQueryMetadata.java:
```diff
@@ -50,32 +50,65 @@ public KeyQueryMetadata(final HostInfo activeHost, final Set<HostInfo> standbyHo
+    /**
+     * Get the store partition corresponding to the key.
+     *
+     * @return store partition number
```
I feel it should be "key partition number" instead of store.

johnthotekat replied:
We are fetching the state store partition number corresponding to the key with this getter and hence that description I believe.
More over, this is a copy paste work, we copied existing getter's with a new name and deprecated the old getter's. 
Any specific reason why you feel that the description should be "key partition number" ?

johnthotekat replied:
@mjsax Please correct me if I'm wrong.

brary replied:
I wrote the old getters & javadocs but now I feel that key partition number better conveys the meaning than store partition number.
The class KeyQueryMetadata contains the information for a key being requested via query. So, the partition number we are returning here is the partition to which the key belongs, it has no relation to any store yet. You will of course use this partition information to fetch the data corresponding to the key from any store but still that doesn't change the fact that this partition number is related to the key rather than the store.
I am fine either way though.

mjsax replied:
IMHO, "key partition number" sound like as if the key would be partitioned, and we get the partition number for the "key partition". But in fact, the store is partitioned and thus I think the current "store partition number" is ok as-is.

brary replied:
Sure, that's fine.

brary commented on streams/src/main/java/org/apache/kafka/streams/KeyQueryMetadata.java:
```diff
@@ -50,32 +50,65 @@ public KeyQueryMetadata(final HostInfo activeHost, final Set<HostInfo> standbyHo
+    }
+
+    /**
+     * Get the store partition corresponding to the key.
```
Get the partition number corresponding to the key.

johnthotekat replied:
Relatable to the comments from line 106 right ?

mjsax commented:
Retest this please.

mjsax commented:
Retest this please.

brary commented:
Thanks @johnthotekat for the PR. LGTM!

johnthotekat committed:
Merge pull request #1 from apache/trunk

Merge apache/kafka to johnthotekat/kafka

Commit:
Merge branches 'KAFKA-10316-renaming-getter-method' and 'trunk' of https://github.com/johnthotekat/kafka into KAFKA-10316-renaming-getter-method

johnthotekat commented:
@mjsax  One of the test failed for me in local, This one - KafkaAdminClientTest#testMetadataRetries. 
It's unrelated to the changes in this PR. 
Rest of the tests PASSED without any issues. 

I guess its the same issue we are addressing with https://issues.apache.org/jira/browse/KAFKA-10311 ? There is also an open PR under this. 
```
org.apache.kafka.clients.admin.KafkaAdminClientTest > testMetadataRetries FAILED
    java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.TimeoutException: Call(callName=describeTopics, deadlineMs=1596638067201, tries=1, nextAllowedTryMs=1596638067302) timed out at 1596638067202 after 1 attempt(s)
        at org.apache.kafka.common.internals.KafkaFutureImpl.wrapAndThrow(KafkaFutureImpl.java:45)
        at org.apache.kafka.common.internals.KafkaFutureImpl.access$000(KafkaFutureImpl.java:32)
        at org.apache.kafka.common.internals.KafkaFutureImpl$SingleWaiter.await(KafkaFutureImpl.java:89)
        at org.apache.kafka.common.internals.KafkaFutureImpl.get(KafkaFutureImpl.java:260)
        at org.apache.kafka.clients.admin.KafkaAdminClientTest.testMetadataRetries(KafkaAdminClientTest.java:997)

        Caused by:
        org.apache.kafka.common.errors.TimeoutException: Call(callName=describeTopics, deadlineMs=1596638067201, tries=1, nextAllowedTryMs=1596638067302) timed out at 1596638067202 after 1 attempt(s)

            Caused by:
            org.apache.kafka.common.errors.TimeoutException: Timed out waiting for a node assignment.
```

mjsax commented:
@johnthotekat Some testa are know to be flaky and we try to work on them to avoid those build issues. It's constant effort to keep the build stable. Will trigger a new Jenkins run and can merge after it passed.

I also saw that you merge changes from `trunk` and pushed an update. As long as there a not merge conflicts, this is not necessary. It also "invalidates" the Jenkins test result and we need to rerun Jenkins. Thus, only push new stuff into the PR if necessary. Otherwise, the merging gets delayed unnecessarily. (Also note, the "Retest this please" comments are not for you -- they trigger Jenkins (if Jenkins cooperates...) to rerun and you can ignore those comments).

mjsax commented:
Retest this please.

mjsax commented:
Retest this please.

mjsax commented:
Retest this please.

mjsax commented:
Ok to test.

mjsax commented:
I'll try again later... Jenkins seems to be on holidays...

mjsax commented:
Retest this please.

johnthotekat commented:
> @johnthotekat Some testa are know to be flaky and we try to work on them to avoid those build issues. It's constant effort to keep the build stable. Will trigger a new Jenkins run and can merge after it passed.
> 
> I also saw that you merge changes from `trunk` and pushed an update. As long as there a not merge conflicts, this is not necessary. It also "invalidates" the Jenkins test result and we need to rerun Jenkins. Thus, only push new stuff into the PR if necessary. Otherwise, the merging gets delayed unnecessarily. (Also note, the "Retest this please" comments are not for you -- they trigger Jenkins (if Jenkins cooperates...) to rerun and you can ignore those comments).

I did a pull to my branch just to make sure that my  branch has all the recent changes (reason behind is the flaky test and I saw the PR for that flaky fix) and that's how that merge from `trunk` came in. I made sure it didn't have any conflicts. I'll make a note of your point. Thanks.

And I really thought you put in the comments "Retest this please". :)

mjsax commented:
Thanks for the KIP and PR @johnthotekat. Merged to `trunk`.

mjsax commented:
One more thing: could you do a second PR to update the docs accordingly? We should list the KIP at lest in `docs/streams/upgrade-guide.html` (there should be already a section for 2.7 release). And maybe also check the actual documentation if we need to do any updates theres, too?

johnthotekat commented:
@mjsax Made the changes and raised PR : https://github.com/apache/kafka/pull/9133 
Checked the actual documentations, we're good here.
